src/events/saga/index.ts

```diff
@@ -78,119 +111,128 @@ export class Saga {
     return this
   }
 
-  onError(handler: (saga, event: EventicleEvent, error: Error) => Promise<void>): Saga {
+  onError(handler: (saga, event: EventicleEvent, error: Error) => Promise<void>): Saga<InstanceData> {
     this.errorHandler = handler
     return this
   }
 }
 
-const SAGAS: Saga[] = []
-
-export async function removeAllNotifyIntents(sagaInstance: SagaInstance): Promise<void> {
-  let notifies = await dataStore().findEntity("system", "saga-notify-intent", {instanceId: sagaInstance.internalData.instanceId})
-
-  await Promise.all(notifies.map(value => dataStore().deleteEntity("system", "saga-notify-intent", value.id)))
-}
+const SAGAS: Saga<any>[] = []
 
 export async function removeAllSagas(): Promise<void> {
   SAGAS.forEach(value => value.streamSubs.forEach(sub => sub.close()))
   SAGAS.length = 0
 }
 
-async function checkNotifyIntents(saga: Saga, event: EventicleEvent) {
+async function checkSagaEventHandlers(saga: Saga<any>, event: EventicleEvent) {
 
-  let notifies: NotifySub[] = (await dataStore().findEntity("system", "saga-notify-intent", {
-    saga: saga.name,
-    eventType: event.type
-  })).map(value => value.content)
+  let handler = saga.eventHandler.get(event.type)
 
-  let matchingNotifies = notifies.filter(value => {
-    if (value && value.hasOwnProperty('filterProp')) {
-      if (event.hasOwnProperty(value.filterProp)) {
-        return event[value.filterProp] == value.filterVal
-      }
-      return event.data[value.filterProp] == value.filterVal
-    } else {
-      return false
-    }
-  })
+  let matcher = handler.config.matchInstance(event)
 
-  await Promise.all(matchingNotifies.map(async value => {
-    await apmJoinEvent(event, saga.name + ":" + event.type, "saga-step-" + saga.name, event.type)
-    await span(event.type, {}, async theSpan => {
-      let instanceData = await dataStore().findEntity("system", "saga-instance", {instanceId: value.instanceId})
-
-      let instance = new SagaInstance(instanceData[0].content, instanceData[0])
-
-      if (theSpan) theSpan.setType("SagaStep")
-      await saga.eventHandler.get(event.type).call(instance, instance, event)
-      instance.internalData.events.push(event)
-      instance.record.content = instance.internalData
-      await dataStore().saveEntity("system", "saga-instance", instance.record)
-      await persistNotificationSubs(saga, instance)
-      if (instance.internalData.ended && !instance.internalData.preserveInstanceData) {
-        await dataStore().deleteEntity("system", "saga-instance", instance.record.id)
-      }
-    })
-    if (matchingNotifies.length > 0) {
+  let query = {
+    saga: saga.name
+  }
+
+  query[matcher.instanceProperty] = matcher.value
+
+  logger.debug("Searching for saga-instance", query)
+
+  let instanceData = (await dataStore().findEntity("system", "saga-instance", query))
+
+  logger.debug("Search results for saga-instance", instanceData)
+
+  if (instanceData.length > 0) {
```

Comment:
could it be that there's more than one match?

Reply:
yes, I think so.   The benefit of the previous method is that its obvious if that happens.  

I'll alter it to call all of the saga instances if this matches.   Makes more sense that way I think